google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -410,6 +455,113 @@ private boolean isAlive() {
+                    clientStream.send(StreamingPullRequest.newBuilder().build());
+                    lastClientPingTime.set(clock.nanoTime());
+                    logger.log(Level.FINEST, "Sent client keepalive ping");
+                  }
+                } finally {
+                  lock.unlock();
+                }
+              } catch (Exception e) {
+                logger.log(Level.FINE, "Error sending client keepalive ping", e);
+              }
+            },
+            CLIENT_PING_INTERVAL.getSeconds(),
+            CLIENT_PING_INTERVAL.getSeconds(),
+            TimeUnit.SECONDS);
+  }
+
+  private void stopClientPinger() {
+    if (pingSchedulerHandle != null) {
+      pingSchedulerHandle.cancel(false);
+      pingSchedulerHandle = null;
+    }
+  }
+
+  private void startServerMonitor() {
+    if (serverMonitorHandle != null) {
+      serverMonitorHandle.cancel(false);
+    }
+
+    Duration checkInterval = Duration.ofSeconds(15);
+    serverMonitorHandle =
+        systemExecutor.scheduleAtFixedRate(
+            () -> {
+              try {
+                if (!isAlive()) {
+                  return;
+                }
+
+                long now = clock.nanoTime();
+                long lastResponse = lastServerResponseTime.get();
+                Duration elapsedSinceResponse = Duration.ofNanos(now - lastResponse);
+
+                if (elapsedSinceResponse.compareTo(SERVER_TIMEOUT_DURATION) <= 0) {
+                  return;
+                }
+
+                long lastPing = lastClientPingTime.get();
+                if (lastPing > lastResponse) {
+                  Duration elapsedSincePing = Duration.ofNanos(now - lastPing);
+                  if (elapsedSincePing.compareTo(SERVER_PING_TIMEOUT_DURATION) <= 0) {
+                    // waiting for response from ping
+                    return;
+                  }
+
+                  logger.log(
+                      Level.WARNING,
+                      "No response from server for {0} seconds, and no response to ping sent {1}"
+                          + " seconds ago. Closing stream.",
+                      new Object[] {
+                        elapsedSinceResponse.getSeconds(), elapsedSincePing.getSeconds()
+                      });
+                } else {
+                  logger.log(
+                      Level.WARNING,
+                      "No response from server for {0} seconds. Closing stream.",
+                      elapsedSinceResponse.getSeconds());
+                }
+
+                lock.lock();
+                try {
+                  if (clientStream != null) {
+                    clientStream.closeSendWithError(
+                        Status.UNAVAILABLE
+                            .withDescription("Keepalive timeout with server")
+                            .asException());
+                  }
+                } finally {
+                  lock.unlock();
+                }
+                stopServerMonitor();
```

Comment:
This might not be necessary since closing the clientStream should trigger `onError` which will also call `stopServerMonitor`.

Reply:
You're right that `clientStream.closeSendWithError()` will eventually trigger `onError` but since it's async there might be a delay between the call to `closeSendWithError()` and the stream actually being terminated. So, there's a possibility that this runs again 15 seconds later before `onError` or `onComplete` runs which would result in calling the `closeSendWithError()` twice. If we call `stopServerMonitor()` here, it will prevent this scenario. I'm not sure if that's a scenario we need to worry about though.